### ISSUE TITLE
Operation timeout expired on iotedge restart

### ISSUE THREAD

**Issue by ddimitrov90** [Azure/iotedge#121]
Another issue that started happening after the Iot Edge feature became generally available.

After initial setup of a new edge device, following the steps in the documentation, once the deployment.json manifest is applied, all custom modules are deployed to the edge device, but fail to start with the following exception:
```
Unhandled Exception: System.AggregateException: One or more errors occurred. (Operation timeout expired.) ---> System.TimeoutException: Operation timeout expired.
   at Microsoft.Azure.Devices.Client.InternalClient.<>c.<ApplyTimeout>b__62_2(Task t)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at ObuCommunication.Program.InitDeviceInternal() in /opt/vsts/work/1/s/modules/ObuCommunication/Program.cs:line 323
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at ObuCommunication.Program.Main(String[] args) in /opt/vsts/work/1/s/modules/ObuCommunication/Program.cs:line 43
```

Basically the module client timeouts at exactly 4 minutes after the start and after the containers are restarted, everything is back to normal. This also happens if I manually restart the Iot Edge runtime, while everything is working. All custom modules are restarted, but they fail to initialize with the timeout exception.

Sometimes it happens to fail for 2 consecutive times, which means that it takes at least 8 minutes to start a module.

Also if I dont wait for the full 4 minutes, but manually do a **docker restart** of my custom module, it is successfully started, so may be it is some racing issue or something is not initialized at startup. 

The modules are deployed on a Ubuntu 16.04, there are no problems with the network connectivity. 

What might be the reason for this?

**Comment by jason-e-gross** [comment]
There seems to be some issue with MQTT as the transport in the module.  Folks are currently suggesting to use AQMP for now.

You WILL still experience a client disconnect exactly 4 mins after it's successfully opened (I am, on every module made so far in both Linux and windows) , with a status and reason of client-closed /disabled, but if you add a connection state handler to your code and just reopen the client, it's good to go.

________________________________

**Comment by mrohera** [comment]
@ddimitrov90 yes this is a known issue we are still working on. Here is the thread where the investigation is in progress. #https://github.com/Azure/azure-iot-sdk-csharp/issues/521

**Comment by ddimitrov90** [comment]
thanks for the info, i’ve just switched the modules to use amqp and with a 20 sec delay, it manages to open the connection successfully

**Comment by mrohera** [comment]
Thanks for letting us know. We are actively looking at the MQTT issue and are committed to resolving this as soon as possible.

**Comment by myagley** [comment]
We believe this issue has been fixed in DotNetty here: https://github.com/Azure/DotNetty/pull/413

We need to cut a new release of DotNetty and update the SDK with this version.

**Comment by myagley** [comment]
Version of 1.18.0 of the C# SDK has been released to nuget.org. Please give it a try. This will address the OperationTimeoutExceptions seen with MQTT.

**Comment by myagley** [comment]
Closing this issue for now as the fix is to upgrade to 1.18.0 of the C# SDK. Please feel free to reopen if there are any questions/concerns.

**Comment by nhuurnink** [comment]
I have exactly this problem when using MQTT. If I switch to AMQP, everything works fine.

I use `<PackageReference Include="Microsoft.Azure.Devices.Client" Version="1.18.0"/>`
